### PR TITLE
Revise CoVe concurreny and progress handling

### PR DESCRIPTION
**Pull request by dg0yt**
The Concurrency unit (plus ProgressObserver) is what might be used for Mapper, too.

### CODE REVIEW

**Comment by dg0yt** [comment]
@mpickering Hopefully this solves the crashes you reported in GH-1470. Maybe you want to test the CI artifacts from this PR?

**Comment by lpechacek** [comment]
I'm still reviewing. I didn't expect such an extensive rewite. :) The code looks pretty good to me so far.

**Comment by dg0yt** [comment]
> I'm still reviewing. I didn't expect such an extensive rewite. :) The code looks pretty good to me so far.

All right. It took some time to unfold  and to prepare for review. Some more comments:

This was my first serious expedition into multithreadying, and Qt adds its own variations. There might still be some issues left.

One thing which is not explicit in the Qt documentation, but what I consider an important lesson learnt:
- The QtConcurrent API uses generic code (templates).
- The QRunnable API uses dynamic polymorphism (virtual methods).
- (QThread is low-level anyway, and it can be used in two different ways.)

CoVe's original Functor API puts polymorphism on top of QtConcurrent's API, thus limiting the generality of that API. I would call that an "impedance mismatch". Staying with the high-level QtConcurrent API, I chose to switch to generic code, too. Now it is possible to use arbitrary function objects, lambdas and functions pointers, as supported by QtConcurrent.

The generic API in connection with `InplaceImage`, which is just a modified copying behaviour for `QImage`, made it possible to separate the concurrency aspect (horizontal stripes) from the core image processing.

I was tempted to implement the progress and cancelation exchange via the `QFuture`. Accessing progress on QFuture is more expensive (mutex) than atomic integers, and there is no way to detect an interruption request from within the job start by `QtConcurrent::run()`. QtConcurrent seems to be designed for running a large number of task: Imagine a queue 100 small jobs, processed by a pool of 4 threads and than progress and cancel on the `QFuture` would work quite well.
`Concurrency::Progress` is the result of learning how to implement light-weight concurrent memory access without running into patterns which are Undefined Behaviour in modern C++.

For the `Concurrency` API I was thinking about making the `ProgressObserver` parameter a mandatory reference instead of a maybe-null pointer, but this would have had more impact on the `Vectorizer` code where functions take an optional pointer, too. Maybe this can be revisited when accepting more changes to the `Vectorizer` API, which could use pure functions in a few more places. But this would have made the rewrite even more substantial, and the optionality of the observer also has some convenience.

File: 3rd-party/cove/libvectorizer/Concurrency.h [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+/**
+ * Waits for a set of background jobs to be finished.
+ * 
+ * This function continuously exchanges progress with the observer and sends the
+ * current thread to sleep for a short break until all jobs are finished.
+ * Note that this function does not run the event loop: This is left to the
+ * observer if needed.
+ */
+template <typename ResultType>
+void waitForFinished(ProgressObserver* observer, JobList<ResultType>& jobs)
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply 3rd-party/cove/libvectorizer/Concurrency.h]: With another template parameter, a merge seems feasible. (These function were not created by copy and paste, but using similar refactoring of higher level functions. I didn't realize they became nearly identical.)

File: 3rd-party/cove/libvectorizer/FIRFilter.cpp [hunk 2]
```diff
@@ -23,8 +23,8 @@
 
 #include <QImage>
 
+#include "libvectorizer/ProgressObserver.h"
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply 3rd-party/cove/libvectorizer/FIRFilter.cpp]: Ack.

File: 3rd-party/cove/libvectorizer/ParallelImageProcessing.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+InplaceImage HorizontalStripes::makeStripe(QImage& original, int scanline, int stripe_height)
+{
+	InplaceImage result(original,
+	                    original.scanLine(scanline),
+	                    original.width(),
+	                    std::min(scanline + stripe_height, original.height()) - scanline,
+	                    original.bytesPerLine(),
+	                    original.format() );
+	if (original.colorCount() > 0)
+		result.setColorTable(original.colorTable());
```
Comment (dg0yt) [inline 3rd-party/cove/libvectorizer/ParallelImageProcessing.cpp]: Obsolete, handled in InplaceImage constructor...

File: 3rd-party/cove/libvectorizer/ProgressObserver.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2005-2019 Libor Pecháček.
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply 3rd-party/cove/libvectorizer/ProgressObserver.h]: The principle is to add every year where you made a substantial change.

File: 3rd-party/cove/app/mainform.cpp [hunk 5]
```diff
@@ -165,7 +165,7 @@ mainForm::mainForm(QWidget* parent, OpenOrienteering::Map* map,
 
 mainForm::~mainForm()
 {
-	if (progressDialog) progressDialog->deleteLater();
+	delete progressDialog;
```
[the remark this answers is not in the document]
Comment (dg0yt) [reply 3rd-party/cove/app/mainform.cpp]: This is gone anyway, in the final version. However, I am confident that this change was correct, i.e. the underlying concurrency problems solved, at the time I commited this change. (And I really wanted the application to break if my assumptions were wrong.)

**Comment by dg0yt** [comment]
> Declaring concurrency support requires adding a member type definition such as `HorizontalStripes` to the functor class - this is quite intrusive and maybe the missed the goal. OTOH it is quite telling. There might be a way to achieve the latter in a less intrusive way.

I still don't see a less intrusive way. But I removed the indirection via `template JobListBuilder`: the `makeJobs` function template is to be implemented directly in the declared concurrency type (`HorizontalStripes`) now.

> Special treatment of `void` results adds the `std::condional_t` statements. It might good enough to simple return the results as for any other type.

Remove the special treatment of `void`.

The template "magic" is now limited to:
-  implementing the SFINAE helper `Concurrency::supported`,
- using it for selecting alternative implementations of `Concurrency::process` via SFINAE on the return type of these functions with `std::enable_if_t` + `Concurrency::supported`,
- "perfect forwarding".

**Pull request merged**